lazy_tensor_core/lazy_tensor_core/core/optimize.py

```diff
@@ -0,0 +1,90 @@
+from lazy_tensor_core import _LAZYC
+import dataclasses
+from typing import List, Dict, Any, Callable
+import copy
+from torch import fx
+import torch
+
+debug = False
+
+@dataclasses.dataclass
+class GraphInputMatcher:
+    tensor_id_to_arg_idx: Dict[int, int]
+    graph_input_tensor_ids: List[int]
+    # there are 2 categories of graph_input_tensors.
+    # Category 1: those whose id are not found in tensor_id_to_arg_idx. These are
+    # most likely const tensors and we can get its content from graph_input_tensors
+    # Category 2: those whose id are found in tensor_id_to_arg_idx. We should get
+    #  the tensor from method arguments
+    graph_input_ivalues: List[Any]
+
+    # get the real graph input tensors
+    def __call__(self, args):
+        real_input = []
+        for tensor_id, traced_ivalue in zip(self.graph_input_tensor_ids, self.graph_input_ivalues):
+            arg_idx = self.tensor_id_to_arg_idx.get(tensor_id, None)
+            if arg_idx is None:
+                inp = traced_ivalue
+            else:
+                inp = args[arg_idx]
+            real_input.append(inp)
+        return real_input
+
+def force_lazy_device(model: fx.GraphModule):
+    """
+    Factory methods in a Fx graph may create tensors for a specific eager devices.
+    If we take no actions, those eager tensors will be mixed with lazy tensors and
+    cause crash. This method overwrite those eager device to lazy device.
+    """
+    def tolazydevice(dev):
+        if isinstance(dev, torch.device):
+            return torch.device("lazy", index=dev.index)
+        return dev
+
+    for nd in model.graph.nodes:
+        nd.args = tuple(tolazydevice(arg) for arg in nd.args)
+        nd.kwargs = {k: tolazydevice(v) for k, v in nd.kwargs.items()}
+    model.recompile()
```

Comment:
what does 'recompile' actually do here?

Reply:
It's needed after the Fx graph is changed to make sure the forward method for the GraphModule is up to date. Here is the doc string of this method just for convenience:
> Recompile this GraphModule from its ``graph`` attribute. This should be
>        called after editing the contained ``graph``, otherwise the generated
>        code of this ``GraphModule`` will be out of date.